### ISSUE TITLE
Calling issues in ppc64le with GCC 14

### ISSUE THREAD

**Issue by Enchufa2** [mpimd-csc/flexiblas#45]
This was reported for GCC downstream in Fedora, and later was reassigned to FlexiBLAS after further inspection by Jakub Jelinek. See https://bugzilla.redhat.com/show_bug.cgi?id=2264712, particularly Jakub's analysis [here](https://bugzilla.redhat.com/show_bug.cgi?id=2264712#c9).

**Comment by Enchufa2** [comment]
Note also that this issue arised from this fix in OpenBLAS: https://github.com/OpenMathLib/OpenBLAS/issues/4475

**Comment by grisuthedragon** [comment]
It seems that GCC 14 changed something in handling the "hidden" arguments in Fortran. Together some inconsitencies in the usage of `CHARACTER` in LAPACK this seems to produce the error. The basic rule I used in generating the wrappers is the following:
- If a function argument `X` is declared as `CHARACTER X` it is a single CHARACTER and not an array or string. Thus no hidden length argument is generated. This is the case in the `ZLARFX` routine. 
- If a function argument `X` is declared as `CHARACTER*LEN` it is assumed to be an array or string and thus the hidden argument is created, for example in `SGESVJ`. 

It seems that this assumption is no longer valid as it is also show in the quick fix. The question is, does this fix break the compatablity with old GCC versions?

If not I can create an upstream fix during the next days.

**Comment by sharkcz** [comment]
http://fedora.danny.cz/ppc/flexiblas-3.4.1-abi.patch fixed `LAPACK-xeigtstz_se2_in` and `LAPACK-xeigtstz_zsb_in` that were failing previously (https://bugzilla.redhat.com/show_bug.cgi?id=2264712#c19), so I think we are on the right track

**Comment by sharkcz** [comment]
It seems to be safe with regard to previous GCC versions. With the mentioned patch I see no issues when building flexiblas on Fedora 39 with GCC 13.

**Comment by jakubjelinek** [comment]
See also https://gcc.gnu.org/PR100799

**Comment by grisuthedragon** [comment]
@jakubjelinek Thanks for pointing this out.

**Comment by jakubjelinek** [comment]
Note, not too old gfortran with -fc-prototypes-external option can dump C prototypes from Fortran external procedures.
For
```
subroutine foo (a, b, c, d, e, f, g, h)
  character a
  integer b, c, d, e, f, g, h
end subroutine
```
I can see even GCC 4.1 (2005-ish) to pass 9 arguments
foo (a, b, c, d, e, f, g, h, _a)
In e.g. GCC 4.5 dumped with more details like
foo (character(kind=1)[1:1] & a, integer(kind=4) & b, integer(kind=4) & c, integer(kind=4) & d, integer(kind=4) & e, integer(kind=4) & f, integer(kind=4) & g, integer(kind=4) & h, integer(kind=4) _a)
to show the actual types (in a mix of Fortran/C++, basic types printed Fortran-ish, but the arrays and references printed C++ish).
That also shows the only known ABI change in there which was in GCC 7+ that flexiblas already takes into account, that the hidden lengths were before 32-bit integers and now are size_t.

**Comment by grisuthedragon** [comment]
Should be fixed in v3.4.2-rc1.

**Comment by sharkcz** [comment]
right, it looks good now - `100% tests passed, 0 tests failed out of 115`
